### ISSUE TITLE
omit entity properties into usecase

### ISSUE THREAD

**Issue by italojs** [herbsjs/gotu#49]
I would like to have a simlilar feature like [ZOD omit feature](https://www.npmjs.com/package/zod#pickomit) 

The idea is to avoid create multi entities when I need a valueobject for my usecase, e.g

``` js

const User = entitity('User', {
id: field(string),
name: field(string)
}

usecase(
    'create user',
    {
      request: User, // I dont want to receive id here
    [...] 
```

My propose is:

``` js
const User = entitity('User', {
id: field(string),
name: field(string)
}

// option #1
usecase(
    'create user',
    {
      request: User.toValueObject() // omit default values: id, createdAt, deleteAt, updatedAt
    [...] 
```

``` js
// option #2
// we can merge it with option 1
usecase(
    'create user',
    {
      request: User.omit([ 'id', 'createdAt', 'myOtherProperty']) 
    [...] 
```

``` js
// option #3
const User = entitity('User', {
id: field(string, { omitable: true } ),
myCustomProp: field(Number, { omitable: true }
name: field(string)
}

usecase(
    'create user',
    {
      request: User.toValueObject() // Omit 'id' and  'myCustomProp'
    [...] 
```

**Comment by jhomarolo** [comment]
I like the idea of a mix a little bit your sugestions:

1. Have the ```myCustomProp: field(Number, { omitable: true } ``` in the entity
2. The entity that have `omitable = true`, wont be exported in the `toValueObject()` method. By default the method `toValueObject()` wont supress any field, unless they have the `omitable = true.`
3. Also have the `omit()` method to explicit wich field you don't want to export.

What do you think?

**Comment by dalssoft** [comment]
If the goal is to have a `request` object, there is this issue for the same goal but with a different approach: https://github.com/herbsjs/buchu/issues/53

**Comment by italojs** [comment]
@dalssoft I think when we use the https://github.com/herbsjs/buchu/issues/53 approach, we create a strong coupling between the glues, I think if we implements a `omitable` fields, we could have a generic feature that can be used in many scenarios.

I liked the @jhomarolo suggestion

**Comment by dalssoft** [comment]
> we create a strong coupling between the glues

How so?

IMO, the suggestions on this issue creates a undesired coupling Entity -> Usecase by creating features on `gotu` to satisfy needs in `buchu`. Remember: Use cases (`buchu`) can/should know about Entities (`gotu`) but the reverse is not true.

We should invert this dependency (Usecase -> Entity), as suggested in https://github.com/herbsjs/buchu/issues/53 , so a usecase can read metadata from a entity.

BTW, it would be ok to create a feature in `gotu` if it would serve other scenarios related to entities. But looks like `.omit`, `omitable` and `toValueObject` serves just for the `request` scenario.

**Comment by italojs** [comment]
discussion moved to https://github.com/herbsjs/buchu/issues/53

**Comment by italojs** [comment]
I think we could to starting just avoiding the ID

``` js

const User = entity('User',{
  id: id(Number),
  name: field(String),
  password: field(String)
}
[...]
User.toValueObject()
/* return new entity
nickName
password
*/

**Comment by jhomarolo** [comment]
@italojs do you think to start with `toValueObject()` without using the `field(Number, { omitable: true }` ?

**Comment by dalssoft** [comment]
1. If this is not related to `usecase` request, whats the use case here?
2. Is it related to thie PR https://github.com/herbsjs/gotu/pull/57 ?

**Comment by italojs** [comment]
i guess the #57 PR is only for metadata, not a feature

**Comment by italojs** [comment]
@jhomarolo I implemented your suggestion https://github.com/herbsjs/gotu/pull/59

**Comment by dalssoft** [comment]
Guys, I don't recommend putting energy on this feature until we have a clear use case for it

**Comment by italojs** [comment]
beside the use in usecase, that will impact the herbs2Gql and herbs2rest, sometime we need to remove the `presence: true` of some property only to validade the received data. e.g.

``` js
const User = entity('User', {
   id: id(Number, { presence: true }),
   name: field(String, { presence: true })),
   email: field(String, { presence: true, validation: { email: true }  })),
   createdAt: field(Date, { presence: true }))
})
 
[...] 

request: {
  name: String,
  email: String
}

[...]
'validate the received data': step( ctx => {
        const user = User.fromJSON(ctx.req)
        if(!user.isValid())  { 
             // I can't do this validation becase the id and createdAt is required
             // to do it works, today I need to remove the "presence: true" from fields id and createdAt, it's not a good ideia
        }
}
```

with the `omitable fields` I avoid this problem:

``` js
request: User.asValueObject()
[...]
'validate the received data': step( ctx => {
        const user = User.fromJSON(ctx.req)
        if(!user.isValid({ avoidOmitableFields: true })){ 
        }
}

**Comment by italojs** [comment]
@dalssoft

**Comment by dalssoft** [comment]
for this use case I can think in several approaches: 

**(1) Check for error return**

Without extra implementation, it would be possible to check the returned error code:

```js
if(!user.isValid()){ 
	if (user.error...) {
		// check if the only error came from `createdAt`, regarding `presence`.
		// we could create a helper function to do that
	}
}
```

**(2) Validation Ignore or Override**

```js
if(!user.isValid({ ignore: { createdAt: "presence"} })){ 
```

or 

```js
if(!user.isValid({ override: { createdAt: { presence: false } } })){ 
```

with that there is no need to have any extra metadata on the entity (as suggested: `{ omitable: true }`) since `omitable` depends on the context, which takes us to the next approuch.

**(3) Validation Contexts**

Their could have be a extra metadata on the entity for validation scenarios, like:

```js
const User = entity('User', {
	id: id(Number, { presence: true }),
	name: field(String, { presence: true })),
	email: field(String, { presence: true, validation: { email: true }  })),
	createdAt: field(Date, { presence: true }))

	create: validation({
		{ override: { createdAt: { presence: false } } }
	}),
})
```

and

```js
if(!user.isValid({ context: 'create' })){ 
```

**Comment by italojs** [comment]
**approach (1)** We cant check only the `presence: true` fields because I could to omit a required field, the `createdAt` is a good example, it's mandatory but I don't need this field when I'm creating the entity 

**approach (2)**
It's not a good DX because I will need aways to specify my fields and when I have maaany fields to ignore, I will create an ugly code

**approach (3)**
it will generate so much code and the DX will decrease, I prefer to implement this `omitable` function in a `utils` folder at a unique time then write many validation functions for many scenarios.

This feature is useful, it's so necessary that exists a lib to do it ([lodash.omit](https://www.geeksforgeeks.org/lodash-_-omit-method/)) and the typescript implements it as a native feature as well, we cant imagine when the developer will use it and it must to be easy to use.

**Comment by dalssoft** [comment]
@italojs lodash's `omit` is a object "clone" method, not a validation mechanism. not related to the use case you brought

**Comment by VictorTBX** [comment]
I'm working on this issue through https://github.com/herbsjs/buchu/issues/53

**Comment by italojs** [comment]
@VictorTBX already have an pr for this https://github.com/herbsjs/gotu/pull/59

**Comment by VictorTBX** [comment]
@italojs I'm following @dalssoft idea commented on https://github.com/herbsjs/buchu/issues/53

Using the approach: `request: request.from(Item, { ignore: ['createdAt'] }),`

**Comment by dalssoft** [comment]
This issue started as a feature request to improve converting a entity to a request object with the possibility to remove some fields. 

This was implemented here: https://github.com/herbsjs/buchu/issues/53

After some discussion, other request appeared: the option to have some partial or custom validation for a entity.

In order to keep the discussion focused, I've created a new issue to discuss this: https://github.com/herbsjs/gotu/issues/70

Given that, I'm closing this issue here and we can continue the discussion in the new issue.
